File: batches/batches/batch_13_different_implementation_same_algorithm/compare.py
"""
Comparison - Different Implementation Same Algorithm (NOT cheating)
Same binary search, different variable names and structure.
"""


def binary_search(arr, target):
    left, right = 0, len(arr) - 1
    while left <= right:
        center = left + (right - left) // 2
        val = arr[center]
        if val == target:
            return center
        if val < target:
            left = center + 1
        else:
            right = center - 1
    return -1


if __name__ == "__main__":
    a = [1, 3, 5, 7, 9]
    print(binary_search(a, 5))
    print(binary_search(a, 4))

File: batches_csharp/batch_13_different_implementation_same_algorithm/compare.cs
using System;

class BinarySearch {
    static int SearchRec(int[] arr, int lo, int hi, int target) {
        if (lo > hi) return -1;
        int mid = lo + (hi - lo) / 2;
        if (arr[mid] == target) return mid;
        if (arr[mid] < target) return SearchRec(arr, mid + 1, hi, target);
        return SearchRec(arr, lo, mid - 1, target);
    }

    static int Search(int[] arr, int target) {
        return SearchRec(arr, 0, arr.Length - 1, target);
    }

    static void Main() {
        int[] a = {1, 3, 5, 7, 9};
        Console.WriteLine(Search(a, 5));
        Console.WriteLine(Search(a, 4));
    }
}
